akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala

```diff
@@ -57,202 +60,216 @@ private[http] final class BodyPartParser(defaultContentType: ContentType,
     if (illegalHeaderWarnings) log.warning(errorInfo.withSummaryPrepended("Illegal multipart header").formatPretty)
   }
 
-  private[this] var output = collection.immutable.Queue.empty[Output] // FIXME this probably is too wasteful
-  private[this] var state: ByteString ⇒ StateResult = tryParseInitialBoundary
-  private[this] var terminated = false
-
-  def warnOnIllegalHeader(errorInfo: ErrorInfo): Unit =
-    if (illegalHeaderWarnings) log.warning(errorInfo.withSummaryPrepended("Illegal multipart header").formatPretty)
-
-  override def onPush(input: ByteString, ctx: Context[Output]): SyncDirective =
-    if (!terminated) {
-      try state(input)
-      catch {
-        case e: ParsingException ⇒ fail(e.info)
-        case NotEnoughDataException ⇒
-          // we are missing a try/catch{continue} wrapper somewhere
-          throw new IllegalStateException("unexpected NotEnoughDataException", NotEnoughDataException)
+  val in = Inlet[ByteString]("BodyPartParser.in")
+  val out = Outlet[BodyPartParser.Output]("BodyPartParser.out")
+
+  override val shape = FlowShape(in, out)
+
+  override def createLogic(attributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+      private var output = collection.immutable.Queue.empty[Output]
```

Comment:
I vote for doing it separately as it needs quite some benchmarking to be sure about the change.
If you'd like to puck that up go for it through, however we'll need benchmarks and profiling data :)